flax/nnx/nn/normalization.py

```diff
@@ -988,3 +1004,200 @@ def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
+    rngs: rng key.
+  """
+
+  def __init__(
+    self,
+    layer_instance: Module,
+    *,
+    n_steps: int = 1,
+    epsilon: float = 1e-12,
+    dtype: tp.Optional[Dtype] = None,
+    param_dtype: Dtype = jnp.float32,
+    error_on_non_matrix: bool = False,
+    update_stats: bool = True,
+    rngs: rnglib.Rngs,
+  ):
+    self.layer_instance = layer_instance
+    self.n_steps = n_steps
+    self.epsilon = epsilon
+    self.dtype = dtype
+    self.param_dtype = param_dtype
+    self.error_on_non_matrix = error_on_non_matrix
+
+    # We define here self.use_running_average attribute to make
+    # .train() and .eval() work. These methods internally change
+    # self.use_running_average to False in train mode and
+    # to True in eval mode
+    # update_stats flag has the opposite logic.
+    self.use_running_average = not update_stats
+
+    # Initialize batch stat variables:
+    state = nnx.state(self.layer_instance)
+
+    self.batch_stats: dict[tuple, nnx.BatchStat] = nnx.data({})
+    def init_batch_stats(path, param):
+      if param.ndim <= 1 or self.n_steps < 1:
+        return
+      elif param.ndim > 2:
+        if self.error_on_non_matrix:
+          raise ValueError(
+            f'Layer instance parameter is {param.ndim}D but error_on_non_matrix is True'
+          )
+        else:
+          param = jnp.reshape(param, (-1, param.shape[-1]))
+
+      path_u = path + ("u", )
+      path_sigma = path + ("sigma", )
+
+      key = rngs.params()
+      self.batch_stats[path_u] = nnx.BatchStat(
+        initializers.normal()(key, (1, param.shape[-1]), self.param_dtype)
+      )
+      key = rngs.params()
+      self.batch_stats[path_sigma] = nnx.BatchStat(
+        initializers.ones(key, (), self.param_dtype)
+      )
+
+    nnx.map_state(init_batch_stats, state)
+
+  def __call__(
+    self,
+    x,
+    update_stats: tp.Optional[bool] = None,
+  ):
+    """Compute the largest singular value of the weights in ``self.layer_instance``
+    using power iteration and normalize the weights using this value before
+    computing the ``__call__`` output.
+
+    Args:
+      x: the input array of the nested layer
+      update_stats: if True, update the internal ``u`` vector and ``sigma``
+        value after computing their updated values using power iteration. This
+        will help the power iteration method approximate the true singular value
+        more accurately over time.
+
+    Returns:
+      Output of the layer using spectral normalized weights.
+    """
+    update_stats = first_from(
+      update_stats,
+      not self.use_running_average,
+      error_msg="""No `update_stats` argument was provided to SpectralNorm
+        as either a __call__ argument or __init__ argument.""",
+    )
+    state = nnx.state(self.layer_instance)
+    state = nnx.map_state(
+      partial(self._spectral_normalize, update_stats=update_stats),
+      state,
+    )
+    nnx.update(self.layer_instance, state)
```

Comment:
We can skip producing the new state by doing an inplace op, for example a for-loop on a flatten state